### PR TITLE
feat: add change detection module

### PR DESCRIPTION
**Pull request by knhash**
# Change Detection

Add a module to hook into a (selfhosted) changedetection.io module - Similar to Releases but for general websites.

Screenshot 2024-05-08 at 15 20 59

Uses the following yaml:
```
- type: changes
  title: "Change Detection"
  token: <x-api-header token here>
  watches:
    - < uuid of the watched website >
        
```

### CODE REVIEW

File: internal/feed/changedetection.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	changeWatches := make(ChangeWatches, 0, len(watches))
+
+	if len(watches) == 0 {
+		return changeWatches, nil
+	}
+
+	requests := make([]*http.Request, len(watches))
+
+	for i, repository := range watches {
+		request, _ := http.NewRequest("GET", fmt.Sprintf("https://changedetection.knhash.in/api/v1/watch/%s", repository), nil)
```
Comment (svilenmarkov) [inline internal/feed/changedetection.go:33]: I think this was meant to be a configurable URL? 😄

File: internal/feed/changedetection.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
+package feed
+
+import (
+	"fmt"
+	"log/slog"
+	"net/http"
+	"time"
+)
+
+type changeDetectionResponseJson struct {
```
Comment (svilenmarkov) [inline internal/feed/changedetection.go:10]: I don't have change detection set up and I can't find any information about this in their API docs, but are there any other useful fields we could include here? Looks a bit plain having just the last changed time in the list.

**Review by svilenmarkov — CHANGES_REQUESTED** [review]
Hey, thanks for contributing! 

Just a small change regarding the request URL and I'd be happy to get it merged.

File: internal/feed/changedetection.go (hunk 1 above)
Comment (knhash) [reply internal/feed/changedetection.go:33]: Ouch, yes, this is meant to be configurable. Will change.

File: internal/feed/changedetection.go (hunk 2 above)
Comment (knhash) [reply internal/feed/changedetection.go:10]: You were right, it is a bit plain. I was thinking of having atleast the diff of _what the changes are_ in there but that didn't look very nice in the dashboard.

This particular api, `watch`, has the following information:
```
{
    "body": "",
    "browser_steps": [],
    "browser_steps_last_error_step": null,
    "check_unique_lines": false,
    "check_count": 54,
    "date_created": 1712990398,
    "consecutive_filter_failures": 0,
    "extract_text": [],
    "extract_title_as_title": false,
    "fetch_backend": "system",
    "fetch_time": 0.191,
    "processor": "text_json_diff",
    "filter_failure_notification_send": true,
    "filter_text_added": true,
    "filter_text_replaced": true,
    "filter_text_removed": true,
    "has_ldjson_price_data": false,
    "track_ldjson_price_data": null,
    "headers": {},
    "ignore_text": [],
    "in_stock": null,
    "in_stock_only": false,
    "include_filters": [],
    "last_checked": 1715162768,
    "last_error": false,
    "last_viewed": 1714130993,
    "method": "GET",
    "notification_alert_count": 5,
    "notification_body": "{{diff}}",
    "notification_format": "System default",
    "notification_muted": false,
    "notification_title": "{{watch_url}}",
    "notification_screenshot": false,
    "notification_urls": [
        "ntfys://knhash:...@ntfy.knhash.in/..."
    ],
    "paused": false,
    "previous_md5": "3abb082cb908e2a7ec44658f2cbaa364",
    "previous_md5_before_filters": "c519ddc918c36938485a4a4b3014d83f",
    "proxy": null,
    "remote_server_reply": "netlify",
    "sort_text_alphabetically": false,
    "subtractive_selectors": [],
    "tag": "",
    "tags": [
        "99ec52cc-3b64-48d8-8432-8aa36c9d236d"
    ],
    "text_should_not_be_present": [],
    "time_between_check": {
        "weeks": null,
        "days": null,
        "hours": 12,
        "minutes": null,
        "seconds": null
    },
    "title": "Uberspace",
    "trigger_text": [],
    "url": "https://lab.uberspace.de/",
    "uuid": "e8b98edd-0e28-472c-85bc-74e936705190",
    "webdriver_delay": null,
    "webdriver_js_execute_code": "",
    "ignore_status_codes": false,
    "save_button": true,
    "last_notification_error": false,
    "content_type": "text/html; charset=utf-8",
    "last_check_status": 200,
    "history_n": 6,
    "last_changed": 1715103086,
    "viewed": false
}
```

Comment (knhash) [reply internal/feed/changedetection.go:10]: I will look into the other APIs to see if there is any other glance-able data. Ideally I would want the click itself to lead to a `diff` section, trying to figure that part out.

**Comment by knhash** [comment]
https://github.com/glanceapp/glance/assets/8710439/49b78de6-7fda-4e87-92ac-30f567a5256c


I have added a `diff` link to the module now.

**Comment by svilenmarkov** [comment]
Sorry it took a while to get this merged and thanks again for contributing!

**Comment by dhanadhan** [comment]
Not sure if the ChangeDetection API allows this, but I thought to ask. Is it possible to show the last history value instead of just a link to the diff? It would convenient to see the last value for those one line checks like stock availability or product prices. 

https://changedetection.io/docs/api_v1/index.html#api-Watch_History-Get_list_of_available_stored_snapshots_for_watch

**Comment by dogukansahil** [comment]
> Ekran.Kayıt.2024-05-09.at.12.05.07.mov 
> `diff`Modüle şimdi bir bağlantı ekledim .

55ydiff:55ydiff: 

It only looks like this for me, can you share a complete YAML for the changedetection part?

**Pull request merged**